src/include/btree_inline.h

```diff
@@ -2209,3 +2230,65 @@ __wt_btcur_skip_page(
     WT_REF_UNLOCK(ref, previous_state);
     return (0);
 }
+
+/*
+ * __wt_btcur_bounds_row_position --
+ *     A unpositioned bounded cursor need to start its cursor next and prev walk from the lower or
+ *     upper bound depending on which direction it is going. This function calls search near to
+ *     position the cursor appropriately.
+ */
+static inline int
+__wt_btcur_bounds_row_position(
+  WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, bool direction, bool *need_walk)
+{
+
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_ITEM *bound;
+    int exact;
+    bool key_out_of_bounds;
+    uint64_t bound_flag, bound_flag_inclusive;
+
+    key_out_of_bounds = false;
+    cursor = &cbt->iface;
+    bound = direction ? &cursor->lower_bound : &cursor->upper_bound;
+    bound_flag = direction ? WT_CURSTD_BOUND_UPPER : WT_CURSTD_BOUND_LOWER;
+    bound_flag_inclusive =
+      direction ? WT_CURSTD_BOUND_LOWER_INCLUSIVE : WT_CURSTD_BOUND_UPPER_INCLUSIVE;
+    exact = 0;
+
+    if (direction)
+        WT_STAT_CONN_DATA_INCR(session, cursor_bounds_next_unpositioned);
+    else
+        WT_STAT_CONN_DATA_INCR(session, cursor_bounds_prev_unpositioned);
+
+    WT_ASSERT(session, WT_DATA_IN_ITEM(bound));
+    __wt_cursor_set_raw_key(cursor, bound);
+    F_SET(cursor, WT_CURSTD_BOUND_ENTRY);
+    ret = cursor->search_near(cursor, &exact);
+    F_CLR(cursor, WT_CURSTD_BOUND_ENTRY);
+    WT_RET(ret);
+
+    /*
+     * FIXME-WT-9324: When search_near_bounded is implemented then remove this. If search near
+     * returns a higher value, ensure it's within the upper bound.
+     */
+    if (exact == 0 && F_ISSET(cursor, bound_flag_inclusive)) {
+        return (0);
+    } else if (direction ? exact > 0 : exact < 0) {
+        /*
+         * If search near returns a key higher than the lower bound, check the returned key against
+         * the upper bound to ensure that the key is within bounds. Else there are no visible
+         * records, return WT_NOTFOUND.
+         */
+        if (F_ISSET(cursor, bound_flag)) {
+            WT_RET(__wt_row_compare_bounds(
+              session, cursor, S2BT(session)->collator, direction, &key_out_of_bounds));
+            if (key_out_of_bounds)
+                return (WT_NOTFOUND);
```

Comment:
when the return is WT_NOTFOUND, are we not supposed to unpin the page i.e. will cursor move to unpositioned state ?

Reply:
That's a good point, thanks for bringing that up. Will address that and wrap the callers of this function in a WT_ERR which will call the cursor->reset().